### ISSUE TITLE
[doc] join with undef values

### ISSUE THREAD

**Issue by hexcoder-** [Perl/perl5#22205]
**Where**

https://perldoc.perl.org/blead/functions/join as of 05/10/2024.

**Description**

**join EXPR,LIST**

1. The documentation does not state how the return value is being built, when LIST contains 'undef' values.

F.e.:
`my $rec = join(':', undef, undef);
`
does return ':' (with warnings).

What is missing in the documentation is that LIST items are stringified so that undefined values will be transformed to empty strings yielding the result of two empty string LIST items.

(In my first reading I overlooked the statement that `join` ALWAYS returns a string. So I assumed a single LIST item of `undef` would result in an `undef` value).

**Comment by Grinnz** [comment]
All string functions interpret their arguments as strings, with the exception of `length` which special-cases undef inputs.

**Comment by hexcoder-** [comment]
Thanks! 
My main investigation was #22207, which had me look up `join`´s documentation.

**Comment by mauke** [comment]
The very first sentence:

> Joins the separate **strings** of LIST into a single **string** with fields separated by the value of EXPR, and returns that new **string**.

Mentions both that list elements are treated as strings and that the return value is a string.

**Comment by hexcoder-** [comment]
I agree. The implication is that `undef`s are not preserved, which I would have liked to be told explicitly.

Ok, if I am the only person to find this description very terse and incomplete regarding "undefined", fell free to reject this issue.

**Comment by guest20** [comment]
@hexcoder- The counter-question is, how could undefs be preserved in a string when all strings are defined?

**Comment by hexcoder-** [comment]
@guest20: when **all** strings are defined, there are obviously **no** undefs to be preserved.
I would modify the question like this "what should join do when there are undefined arguments?"

The motivation for this issue/question came from my "code review" of https://github.com/Perl/perl5/blob/8cacb84717058962131a5d2a3180c34d514f2b05/lib/perl5db.pl#L2494
and the following line
https://github.com/Perl/perl5/blob/8cacb84717058962131a5d2a3180c34d514f2b05/lib/perl5db.pl#L2495

where the `defined` operator is applied to the result of a `join`.

**Comment by Grinnz** [comment]
I think you missed the point of the question: strings are by definition defined, not in the argument list but in all of Perl. So when undef is passed to a function that takes strings, it is always interpreted as the empty string with a warning.
